src/gamestate.ts

```diff
@@ -33,17 +35,145 @@ export class GameState {
 		this.attack_direction_event = new AttackDirectionEvent();
 		this.tears = new Array<Tear>();
 
+		this.joysticks = new Array<Joystick>();
+
 		this.jays = new Jays();
 		document.onkeyup = event => this.key_up(event.key);
 		document.onkeydown = event => this.key_down(event.key);
-		// });
+
+		document.ontouchstart = event => this.touch_start(event.touches);
+		document.ontouchend = event => this.touch_end(event.touches);
+		document.ontouchmove = event => { this.touches = event.touches; };
 	}
 
-	public key_down(keyName: string): void {
+	public touch_start(touches: TouchList): void {
+		const touches_array = TouchHelper.touchlist_to_id_array(this.touches);
+		const new_touches_array = TouchHelper.touchlist_to_id_array(touches);
+		const new_touch = TouchHelper.get_touch_by_identifier(touches, ArrayUtil.diff(new_touches_array, touches_array)[0]);
 
-		const direction = key_mapper.current_keyboard.get(keyName);
+		this.touches = touches;
+		const touch_x = new_touch.pageX;
+		const touch_y = new_touch.pageY;
+		if (touch_x < (window_W / 2)) {
+			if (!this.joysticks.find(joystick => joystick.id === "LEFT")) {
+				this.joysticks.push(new Joystick("LEFT", new Point(touch_x, touch_y), 40,
+					new Point(touch_x, touch_y), 20, new_touch.identifier));
+			}
+		}
+		else {
+			if (!this.joysticks.find(joystick => joystick.id === "RIGHT")) {
+				this.joysticks.push(new Joystick("RIGHT", new Point(touch_x, touch_y), 40,
+					new Point(touch_x, touch_y), 20, new_touch.identifier));
+			}
+		}
+	}
+
+	public touch_end(touches: TouchList): void {
+		const touches_array = TouchHelper.touchlist_to_id_array(this.touches);
+		const new_touches_array = TouchHelper.touchlist_to_id_array(touches);
+		const touches_removed = ArrayUtil.diff(touches_array, new_touches_array);
+
+		const joysticks_to_remove = new Array<Joystick>();
+
+		touches_removed.forEach(touch_identifier => {
+			const joystick_to_remove = this.joysticks.find(joystick => joystick.touch_identifier === touch_identifier);
+			if (joystick_to_remove != null) {
+				joysticks_to_remove.push(joystick_to_remove);
+			}
+		});
+
+		this.touches = touches;
+
+		joysticks_to_remove.forEach(joystick => {
+			joystick.div_zone.remove();
+			joystick.div_controller.remove();
+			ArrayUtil.remove_from_array(this.joysticks, joystick);
+			switch (joystick.id) {
+				case "LEFT":
+					this.remove_direction_event(Direction.LEFT);
+					this.remove_direction_event(Direction.RIGHT);
+					this.remove_direction_event(Direction.DOWN);
+					this.remove_direction_event(Direction.UP);
+					break;
+				case "RIGHT":
+					this.remove_attack_event(Direction.LEFT);
+					this.remove_attack_event(Direction.RIGHT);
+					this.remove_attack_event(Direction.DOWN);
+					this.remove_attack_event(Direction.UP);
+					break;
+			}
+		});
+	}
+
+	public touch_move(): void {
+		if (this.joysticks.length > 0) {
```

Comment:
Faut inverser la condition et faire un `return;`, ça te fais un nesting énorme

`if > for > if > switch > if > else`

MOMS SPAGHETTI

Reply:
C'est ta barbe les spaghettis